src/NHSD.BuyingCatalogue.Identity.Api/Controllers/AccountController.cs

```diff
@@ -151,7 +154,28 @@ public IActionResult ResetPassword(ResetPasswordViewModel viewModel)
                 return View(viewModel);
             }
 
-            return RedirectToAction(nameof(ResetPasswordConfirmation));
+            var res = await _passwordService.ResetPasswordAsync(viewModel.Email, viewModel.Token, viewModel.Password);
+            if (res.Succeeded)
+            {
+                return RedirectToAction(nameof(ResetPasswordConfirmation));
+            }
+
+            var invalidPasswordError = res.Errors.FirstOrDefault(x => x.Code == PasswordValidator.InvalidPasswordCode);
+            if (invalidPasswordError != null)
+            {
+                ModelState.AddModelError(nameof(ResetPasswordViewModel.Password), invalidPasswordError.Description);
+                return View(viewModel);
+            }
+
+            var invalidTokenError = res.Errors.FirstOrDefault(x => x.Code == "InvalidToken");
+            if (invalidTokenError != null)
+            {
+                return RedirectToAction(nameof(ResetPasswordExpired));
+            }
+
+            throw new Exception(
```

Comment:
Can we throw a more specific exception type?

Reply:
I was looking for more specific exception types and couldn't find one I was happy with. Open to suggestions! Or we could roll our own